### DISCUSSION TITLE
[SOLVED] [BUG] Issues with Add-PnPListItem and Set-PnPListItem for Choice fields with PnPBatch

### DISCUSSION THREAD

**Discussion by bGlasson** [pnp/powershell#2427] [General]
### Reporting an Issue or Missing Feature
Adding/updating Choice fields that accept multiple values is breaks when is used.

Correctly formatted strings (length =n) submitted for Add-PnPListItem and Set-PnPListItem via PnPBatch to add/update a Choice field that allows multiple values are split into n single character values.

### Expected behavior 
string of "value1;#value2" submitted to add/update a mulit-choice field should result in a record with 2 values in that field:
  _v1, v2_

### Actual behavior 
record contains 6 values:
_v,1,;,#,v,2_
 

### Steps to reproduce behavior
Sample repro code 

    `#Connect to PnP Online
    Connect-PnPOnline -Url $SiteURL -UseWebLogin
    
    #batch it
    
    $batch = New-PnPBatch
    
    Add-PnPListItem -List $ListName -Values @{
    
    #Set-PnPListItem -List $ListName -Identity 33 -Values @{
        "Title" = "hardcode title";
        "scopeRef" = "Blue;#Green";
        "dateTest" = "21/09/2022"
        } -Batch $batch
    
    #run it
    Invoke-PnPBatch -Batch $batch `

Results with/without batching:
![image](https://user-images.githubusercontent.com/114933082/194046327-32ddbd79-f7c4-4657-b78d-ec5e50ae4491.png)

### What is the version of the Cmdlet module you are running?
1.11.0

### Which operating system/environment are you running PnP PowerShell on?
- [X] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by gautamdsheth** [comment]
You needn't specify `;#`, we handle that for you.

Changing it to `"scopeRef" = "Blue","Green"` should work.

**Comment by bGlasson** [comment]
First - thanks for prompt response. These tools close major gaps and are seriously useful - thanks 🙇 

If I hardcode a string inside the .ps1, that works (though formatting doesn't seem to match [the documentation here](https://pnp.github.io/powershell/cmdlets/Add-PnPListItem.html#-values)?). 

If I try to read the strings from a CSV for a bulk update, _value1;#value2_ and _value1,value2_ both fail 😢 

```
$batch = New-PnPBatch
 
foreach($item in $csv) 
{
Add-PnPListItem -List $ListName -Values @{

#Set-PnPListItem -List $ListName -Identity 33 -Values @{
    "Title" = $item."Title";
    "scopeRef" = $item."scopeRef";
    "dateTest" = $item."dateTest";
    } -Batch $batch
    }
#run it
Invoke-PnPBatch -Batch $batch 
```

CSV (UTF-8 created/saved in excel):
![image](https://user-images.githubusercontent.com/114933082/194058884-fe8a3222-68b8-4eb2-a3db-5dde998e3766.png)

Results in:
![image](https://user-images.githubusercontent.com/114933082/194059045-0fd7c59d-f9de-4dbe-aadf-f4361cd803e3.png)

**Comment by gautamdsheth** [comment]
You will need to process your CSV data to make it work.
In your logic of importing from CSV, maybe splitting the values and storing them as string arrays will  work.

So, `blue,green` will have to sent as `"blue","green"`. 
So, you need to modify this line `$item."scopeRef";` , read the value and prepare it as comma separated string.

Am moving this to discussion as it is not an issue with the module.
Will update the docs as well to remove the confusion.

**Reply by bGlasson** [reply]
Think I'm missing something here - apologies if I'm being dense. Below is my working step by step. Struggling to make PnPBatch work with anything other than a hardcoded value - can't even find the right string tweak from the CSV for a single value to work. 

Would really appreciate it if someone spots (and shares! 😅) the issue in the string tweaking below... or could share a simple working example of PnPBatch + Add-PnPListItem + CSV for a multi value choice field. I'm stumped! 


If I hardcode _"blue","green"_ it's fine:
![image](https://user-images.githubusercontent.com/114933082/194374901-a0ff65db-9bce-4ee3-b0e7-33e11bc8d06b.png)

If I use the CSV and try to convert _blue,green_ to the same string, I get odd results
Input CSV:
![image](https://user-images.githubusercontent.com/114933082/194378850-9abc5c59-8d71-4256-bf78-2878d832812c.png)


replacing , with "," 
![image](https://user-images.githubusercontent.com/114933082/194375244-d9bf11b7-c5f2-4ee8-9415-e2c4aa9d2bf8.png)

gives
![image](https://user-images.githubusercontent.com/114933082/194375358-7cb65ff5-e975-47c6-a227-c383dde3276d.png)

adding extra quotes to the outside, incase that helps:
![image](https://user-images.githubusercontent.com/114933082/194376265-71b746d1-4277-40d6-a442-02fd57f5376e.png)

gives
![image](https://user-images.githubusercontent.com/114933082/194376392-8fda1de7-61d8-4d61-9b1c-6fe3947fee1a.png)

I added the write-host above to make sure the CSV wasn't encoded/read in a strange way, but it looks as I'd expected:
![image](https://user-images.githubusercontent.com/114933082/194376787-f2867c1b-a4d4-4a8c-ba9e-cfd14fed5cff.png)


My understanding was PnPBatch is supposed to stack "normal" inputs (e.g. stuff that works fine unbatched) with Add-PnPListItem etc for more efficient execution. For me at least, it seems to alter how the process works and I'm not sure how to compensate for it.

The same csv without the batching or extra string fiddling:
![image](https://user-images.githubusercontent.com/114933082/194378783-1208626a-570f-4d36-a854-a678f2c2e083.png)
![image](https://user-images.githubusercontent.com/114933082/194379058-51153bf9-985b-4e68-a981-7f142a5dbf15.png)

Without PnPBatch, the strings pass through as I'd expect. If I saw the , for ;# in either the code or the csv, it gives me results as expected:
`"scopeRef" =  $item."scopeRef".Replace(",",";#") ;`
![image](https://user-images.githubusercontent.com/114933082/194379414-0e961f82-d603-4808-aa79-a3ad7c44ec57.png)

**Reply by bGlasson** [reply]
Thanks to my colleague _JorgeSch_  for dragging me to understanding this.

PnPBatch seems to cause (ambiguous?) strings in the csv file to be interpreted differently, resulting in the single character splits. My attempts to fix this by delimiting the string differently were never going to work...  the solution is to pass an explicit array of strings for these choice fields - simply adding a .split() does the job.

Working example:
```
Add-PnPListItem -List $ListName -Values @{
    "Title" = $item."Title";
    "scopeRef" =  ($item."scopeRef").split(",");
```

![image](https://user-images.githubusercontent.com/114933082/195828203-380867c1-6227-41ec-9831-b66d2b077042.png)

**Comment by SpeedmaxX** [comment]
Hi, Error seems still to exist. I only could solve this for me by leaving out the batch.
